src/self_driving/planning/pilot_util.cpp

```diff
@@ -419,4 +425,233 @@ void PilotUtil::GroupFeaturesByOU(
+      table_oid = accessor->GetTableOid(insert_statement->GetInsertionTable()->GetTableName());
+      // Number of values is the number of rows inserted
+      num_row_delta = insert_statement->GetValues()->size();
+    }
+
+    if (statement->GetType() == parser::StatementType::DELETE) {
+      auto delete_statement = reinterpret_cast<parser::DeleteStatement *>(statement.Get());
+      table_oid = accessor->GetTableOid(delete_statement->GetDeletionTable()->GetTableName());
+      if (delete_statement->GetDeleteCondition() != nullptr) {
+        // TODO(lin): right now assuming only deleting one row when there's a condition. Need cardinality estimation
+        //  to get a more accurate estimation
+        num_row_delta = -1;
+      } else {
+        if (table_sizes.find(table_oid) != table_sizes.end())
+          // Delete all the tuples when without condition
+          num_row_delta = -table_sizes[table_oid];
+      }
+    }
+    if (num_row_delta != 0 && table_sizes.find(table_oid) != table_sizes.end())
+      query_row_changes[query_id] = std::make_pair(table_oid, num_row_delta);
+  }
+  txn_manager->Abort(txn);
+
+  // Compute table size ratios given the workload forecast
+  for (uint64_t idx = 0; idx < forecast->GetNumberOfSegments(); ++idx) {
+    // Get the table num row changes in this segment
+    auto &id_to_num_exec = forecast->GetSegmentByIndex(idx).GetIdToNumexec();
+    std::unordered_map<catalog::table_oid_t, double> table_size_deltas;
+    for (const auto &[query_id, table_id_to_delta] : query_row_changes) {
+      auto table_id = table_id_to_delta.first;
+      if (table_size_deltas.find(table_id) == table_size_deltas.end()) table_size_deltas[table_id] = 0;
+      table_size_deltas[table_id] += table_id_to_delta.second * id_to_num_exec.at(query_id);
+    }
+
+    // Calculate the table size change ratio for this segment
+    for (const auto &[table_id, size_delta] : table_size_deltas) {
+      double new_table_size = size_delta + static_cast<int64_t>(table_sizes[table_id]);
+      if (new_table_size < 0)
+        memory_info->segment_table_size_ratios_[idx][table_id] = 0;
+      else
+        memory_info->segment_table_size_ratios_[idx][table_id] = new_table_size / table_sizes[table_id];
+    }
+  }
+}
+
+void PilotUtil::ComputeTableIndexSizes(common::ManagedPointer<transaction::TransactionManager> txn_manager,
+                                       common::ManagedPointer<catalog::Catalog> catalog,
+                                       pilot::MemoryInfo *memory_info) {
+  catalog::db_oid_t max_db_oid = catalog->GetNextOid();
+  auto txn = txn_manager->BeginTransaction();
+  // Traverse all databases to find the info (since we don't know which table belongs to which database)
+  for (auto db_oid = catalog::db_oid_t(1); db_oid < max_db_oid; db_oid++) {
+    auto accessor = catalog->GetAccessor(common::ManagedPointer(txn), db_oid, DISABLED);
+    if (accessor == nullptr) continue;
```

Comment:
This should probably be replaced with a method on the `Catalog` that takes a transaction and returns all current database OIDs (and names?).

Reply:
This is actually not needed anymore once we create transactions for every database at the beginning of the planning and pass them along.